planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+* Finer-grained manipulation of the `IIdentityMap` 
+
+
+## Unit Testing Approach
+
+If we introduce the runtime code generation back into Marten, that's unfortunately a non-trivial "cold start" testing issue. To soften that, I suggest we get a lot more aggressive with reusable [xUnit.Net class fixtures](https://xunit.net/docs/shared-context) between tests to reuse generated code between tests, cut way down on the sheer number of database calls by not having to frequently check the schema configuration, and other `DocumentStore` overhead. 
+
+A couple other points about this:
+
+* We need to create more unique document types so we're not having to use different configurations for the same document type. This would enable more reuse inside the testing runtime
+* Be aggressive with separate schemas for different configurations
+* We could possibly turn on xUnit.net parallel test running to speed up the test cycles
+
+
+## Document Metadata
+
+* From the feedback on GitHub, it sounds like the desire to extend the existing metadata to tracking data like correlation identifiers, transaction ids, user ids, etc. To make this data
+easy to query on, I would prefer that this data be separate columns in the underlying storage 
+
+* Use the configuration and tests from [pull request for Document Metadata](https://github.com/JasperFx/marten/pull/1364), but use the Lamar-backed dynamic code generation from the previous section to pull this off.
+
+* I strongly suggest using a new dynamic codegen model for the `ISelector` objects that would be responsible for putting Marten's own document metadata like `IsDeleted` or `TenantId` or `Version` onto the resolved objects (but that falls apart if we have to use private setters)
+* I think we could expand the document metadata to allow for user defined properties like "user id" or "transaction id" much the same way we'll do for the EventStore metadata. We'd need to think about how we extend the document tables and how metadata is attached to a document session
+
+My thought is to designate one (or maybe a few?) .Net type as the "metadata type" for your application like maybe this one:
+
+```
+    public class MyMetadata
+    {
+        public Guid CorrelationId { get; set; }
+        public string UserId { get; set; }
+    }
+```
+Maybe that gets added to the `StoreOptions` something like:
+
+```
+var store = DocumentStore.For(x => {
+    // other stuff
+
+    // This would direct Marten to add extra columns to
+    // the documents and events for the metadata properties
+    // on the MyMetadata type.
+
+    // This would probably be a fluent interface to optionally fine tune
+    // the storage and applicability -- i.e., to all documents, to events, etc.
+    x.UseMetadataType<MyMetadata>();
+});
+```
+
+Then at runtime, you'd do something like:
+
+```
+session.UseMetadata<MyMetadta>(metadata);
```

Comment:
Setting up it on the store configuration level is straightforward for me, but why would we like to set it on the session level? If we'd like to go with the elastic configuration then wouldn't it be good to set up it per document type or even document instance?

Reply:
We'd definitely do it document by document type for built in metadata, but would it really need to vary for things like causation id / user id kind of tracking?

Reply:
Currently, it's hard to say for me. I think that we dive a bit more on how others are doing that. Some food for thought:
- https://eventstore.com/docs/server/metadata-and-reserved-names/index.html
- https://railseventstore.org/docs/correlation_causation/
My colleague is the Rails Event Store maintainer. I'll try to make some inquiry ;)

Reply:
Missed this. You'd have to:

1. Do something on `StoreOptions` to declare any kind of custom metadata type so that we can add custom fields on the document or event storage for things like correlation id or causation id.
1. Set the active metadata values at the session level